downstairs/src/lib.rs

```diff
@@ -1074,7 +1156,11 @@ where
                     }
                     Some(m) => {
                         let mut ds = ads.lock().await;
-                        ds.on_negotiation_step(m, &mut state, &mut fw).await?;
+                        ds.on_negotiation_step(
```

Comment:
>  Is that going to cause any chances for deadlock? 

I don't think there a possibility for deadlock, because all of the channel are unbounded.

> I'm assuming we are moving toward a one task that does not need as many locks with this refactor?

Yes, in the one-task world, the `Downstairs` is not behind a lock **at all**; everyone communicates with it by message-passing.